be/src/storage/table_reader.h

```diff
@@ -0,0 +1,104 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Inc.
+
+#pragma once
+
+#include <future>
+#include <utility>
+
+#include "column/row.h"
+#include "storage/chunk_iterator.h"
+#include "storage/vectorized_column_predicate.h"
+
+namespace starrocks {
+
+using ColumnPredicate = vectorized::ColumnPredicate;
+using ChunkIteratorPtr = vectorized::ChunkIteratorPtr;
+
+// Parameters used to create a TableReader.
+struct TableReaderParams {
+    // Version of data to read
+    Version version;
+
+    // Schema for the key
+    vectorized::VectorizedSchema key_schema;
+
+    // Schema for the value
+    vectorized::VectorizedSchema value_schema;
+
+    // Chunk size used to read data
+    int chunk_size = 1024;
+
+    // Common predicates for all read requests. This may be useful
+    // when reading data remotely. Some data can be filtered
+    // in the remote and reduce network cost.
+    std::vector<const vectorized::ColumnPredicate*> common_predicates;
+};
+
+// Options used to control the behaviour of a single read request.
+struct ReadOption {
+    // Whether to return the result in the sorted order of the key.
+    // This may be used in the prefix scan where there may be many
+    // keys under a given prefix.
+    bool strict_order = false;
+
+    // Whether to return the result in the reverse order of the key.
+    // This option is available when strict_order is true.
+    bool reverse_order = false;
+
+    // Maximum number of rows to return. -1 indicates there is
+    // no limit, and all rows need to return
+    int row_count_limit = -1;
+
+    // Custom predicates for a read request
+    std::vector<const vectorized::ColumnPredicate*> predicates;
+};
+
+// Table reader will read data in the key-value style. The key and value contains
+// one or multiple columns which are defined by TableReaderParams#key_schema and
+// TableReaderParams#value_schema respectively. The key may not have the unique
+// constraint, that's, multiple values may be returned for a key. You can also read
+// data by the prefix of a key. Table reader can read data from local and remote
+// tablets which is up to where is the data, and whether the implementation supports
+// to read remotely.
+class TableReader {
+public:
+    explicit TableReader(const TableReaderParams& params) : _reader_params(params) {}
+
+    // Get the values under the given key, and organize the result in Chunks. The schema
+    // of the returned chunk is defined by TableReaderParams#value_schema.
+    virtual StatusOr<ChunkIteratorPtr> get(const Row& key, const ReadOption& read_option) = 0;
```

Comment:
get/multiget return StatusOr<ChunkPtr> instead of StatusOr<ChunkIteratorPtr> ?

Reply:
The concern is that a `Chunk` can not hold all of data in some cases, eg., there are many values under a given key. We can make the chunk size bigger, but it may be not friendly for memory allocation.

Reply:
if we just `get` a given key, then will return a chunk with one row which wouldn't too big, otherwise `ChunkIteratorPtr` have the same problem. If we use `multiget`, we can control the rows size that we want to get, to avoid cost too much memory.

Reply:
Maybe not just one row. For lookup join, there maybe many rows under a key.